lib/oauth-db.js

```diff
@@ -38,65 +58,111 @@ OAuthDB.prototype.generateAuthCode = function(clientId, userId, scopes, expiresA
 
     var authCode = random.toString('hex');
 
-    this.authCodes[authCode] = {
-      client_id: clientId,
-      user_id: userId,
-      scopes: scopes,
-      expires_at: expiresAt
-    };
-
-    callback(null, authCode);
+    var saltedCode = crypto.createHash('sha256')
+      .update(authCode + secret)
+      .digest('hex');
+
+    executeQuery(queries.create.authCode, [
+      saltedCode,
+      clientId,
+      userId,
+      JSON.stringify(scopes),
+      expiresAt
+    ], function(err) {
+      if ( err ) {
+        return callback(err);
+      }
+
+      process.nextTick(function() {
+        callback(null, authCode);
+      });
+    });
   }.bind(this));
 };
 
 OAuthDB.prototype.verifyAuthCode = function(authCode, clientId, callback) {
-  if ( !this.authCodes[authCode] ) {
-    return process.nextTick(function() {
-      callback(Boom.forbidden('invalid auth code'));
-    });
-  }
-
-  if ( this.authCodes[authCode].client_id !== clientId ) {
-    return process.nextTick(function() {
-      callback(Boom.forbidden('invalid client id'));
+  var saltedCode = crypto.createHash('sha256')
+    .update(authCode + secret)
+    .digest('hex');
+
+  executeQuery(queries.get.authCode, [saltedCode],  function(err, code) {
+    if ( err ) {
+      return process.nextTick(function() {
+        callback(err);
+      });
+    }
+
+    if ( !code ) {
+      return process.nextTick(function() {
+        callback(Boom.forbidden('invalid auth code'));
+      });
+    }
+
+    if ( code.client_id !== clientId ) {
+      return process.nextTick(function() {
+        callback(Boom.forbidden('invalid client id'));
+      });
+    }
+
+    if ( code.expires_at <= Date.now() ) {
+      return process.nextTick(function() {
+        callback(Boom.forbidden('auth code expired'));
+      });
+    }
+
+    process.nextTick(function() {
```

Comment:
You should delete the authCode or update the used column after fetching it w/o error